web-api/src/storage/postgres.rs

```diff
@@ -382,56 +446,96 @@ impl Database {
-                            snippet,
-                            is_summarized,
-                            properties,
-                            tags,
-                            embedding,
-                            is_candidate: true,
-                        })
-                    })
-                    .fetch_all(&mut tx)
-                    .await?,
-            );
-        }
+        let needs_ingestion =
+            Self::set_is_candidate_and_return_for_ingestion(&mut tx, ingestable.iter().copied())
+                .await?;
 
         tx.commit().await?;
 
-        let failed = (ingested.len() < ingestable.len())
+        let failed = (needs_ingestion.len() < ingestable.len())
             .then(|| {
-                for document in &ingested {
+                for document in &needs_ingestion {
                     ingestable.remove(&document.id);
                 }
                 ingestable.into_iter().cloned().collect()
             })
             .unwrap_or_default();
 
-        Ok((unchanged, ingested, failed))
+        Ok((unchanged, needs_ingestion, failed))
+    }
+
+    async fn set_is_candidate_and_return_for_ingestion(
+        tx: &mut Transaction<'_, Postgres>,
+        ids: impl ExactSizeIterator<Item = &DocumentId> + Clone,
+    ) -> Result<Vec<DocumentForIngestion>, Error> {
+        let mut builder = QueryBuilder::new(
+            "SELECT document_id, snippet_idx, snippet, embedding
+            FROM snippet
+            WHERE document_id IN ",
+        );
+        let mut snippets = HashMap::<_, BTreeMap<_, _>>::new();
+        let mut chunks = IterAsTuple::chunks(Self::BIND_LIMIT, ids.clone());
+        while let Some(ids) = chunks.next() {
+            let mut chunk = builder
+                .reset()
+                .push_tuple(ids)
+                .build_query_as::<SqlSnippet>()
+                .fetch(&mut *tx);
+
+            while let Some(snippet) = chunk.next().await.transpose()? {
+                snippets.entry(snippet.document_id).or_default().insert(
+                    u32::from(snippet.snippet_idx),
+                    (snippet.snippet, snippet.embedding),
+                );
+            }
+        }
+
+        let mut builder = QueryBuilder::new(
+            "UPDATE document
+            SET is_candidate = TRUE
+            WHERE document_id IN ",
+        );
+        let mut needs_ingestion = Vec::with_capacity(ids.len());
+        let mut chunks = IterAsTuple::chunks(Self::BIND_LIMIT, ids);
+        while let Some(ids) = chunks.next() {
+            let mut rows = builder
+                .reset()
+                .push_tuple(ids)
+                .push(" RETURNING document_id, preprocessing_step, properties, tags;")
+                .build()
+                .fetch(&mut *tx);
+
+            while let Some(row) = rows.next().await.transpose()? {
+                let document_id = row.try_get("document_id")?;
+                //Hint: We currently assume there are no gaps
```

Comment:
❓ by "no gaps" do you mean that we assume there is at least one snippet in the snippet table for every document id in the document table?

Reply:
I mean that if there are 10 snippets for a specific document then their `snippet_idx` are 0..10, clarified it.